src/prelude/timers.mo

```diff
@@ -0,0 +1,149 @@
+// default timer mechanism implementation
+// fundamental node invariant: max_exp ante <= expire <= min_exp dopo
+// corollary: if expire == 0 then the ante is completely expired
+//
+type @Node = { var expire : Nat64; id : Nat; delay : ?Nat64; job : () -> async (); ante : ?@Node; dopo : ?@Node };
+var @timers : ?@Node = null;
+func @prune(n : ?@Node) : ?@Node = switch n {
+    case null null;
+    case (?n) {
+        if (n.expire == 0) {
+            @prune(n.dopo) // by corollary
+        } else {
+            ?{ n with ante = @prune(n.ante); dopo = @prune(n.dopo) }
+        }
+    }
+};
+
+func @nextExpiration(n : ?@Node) : Nat64 = switch n {
+    case null 0;
+    case (?n) {
+        var exp = @nextExpiration(n.ante); // TODO: use the corollary for expire == 0
+        if (exp == 0) {
+            exp := n.expire;
+            if (exp == 0) {
+                exp := @nextExpiration(n.dopo)
+            }
+        };
+        exp
+    }
+};
+
+// Function called by backend to run eligible timed actions.
+// DO NOT RENAME without modifying compilation.
+func @timer_helper() : async () {
+    func Array_init<T>(len : Nat,  x : T) : [var T] {
+        (prim "Array.init" : <T>(Nat, T) -> [var T])<T>(len, x)
+    };
+    let now = (prim "time" : () -> Nat64)();
+    let exp = @nextExpiration @timers;
+    let prev = (prim "global_timer_set" : Nat64 -> Nat64) exp;
+
+    if (exp == 0) {
+        return
+    };
+
+    var gathered = 0;
+    let thunks = Array_init<?(() -> async ())>(10, null); // we want max 10
+
+    func gatherExpired(n : ?@Node) = switch n {
+        case null ();
+        case (?n) {
+            gatherExpired(n.ante);
+            if (n.expire > 0 and n.expire <= now and gathered < thunks.size()) {
+                thunks[gathered] := ?(n.job);
+                switch (n.delay) {
+                    case (?delay) {
+                        // re-add the node
+                        let expire = n.expire + delay;
+                        // N.B. insert only works on pruned nodes
+                        func reinsert(m : ?@Node) : @Node = switch m {
+                            case null ({ n with var expire; ante = null; dopo = null });
+                            case (?m) {
+                                assert m.expire != 0;
+                                if (expire < m.expire) ({ m with ante = ?reinsert(m.ante) })
+                                else ({ m with dopo = ?reinsert(m.dopo) })
```

Comment:
```suggestion
                                if (expire < m.expire) {
                                  m.before :=  ?reinsert(m.before) 
                                }
                                else { 
                                  m.after := ?reinsert(m.after)
                                }
```
With the imperative (renamed) version, you could just insert in place, without re-allocating (I think).